Adding CredSSP with Early User Auth support (sec:ext)

hlefebvre opened the pull request:
Recently I wasn't able to connect to my work servers using FreeRDP.

After digging I found that without using /sec:ext the server was closing the connection.

But with the command line switch it wasn't working, I found out that HYBRID_EX wasn't handled in nego.c.

Afterward I found out that we were receiving 4 bytes between the end of NLA and the MCS connect, I found out that it was the Early User Auth Result PDU (https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpbcgr/d0e560a3-25cb-4563-8bdc-6c4cc625bbfc).

I don't know RDP protocol neither the FreeRDP source code but I attempted to add support the best I can, trying to respect the way the code is architectured and using clang_format on the modified sources.

With this patch I can now connect to my work servers again, if it can help other people then I'm happy to share the patch.

akallabeth commented:
@freerdp-bot test please

akallabeth commented on libfreerdp/core/nla.c:
```diff
@@ -1552,63 +1580,80 @@ int nla_recv_pdu(rdpNla* nla, wStream* s)
 	{
 		UINT32 code;
+		Stream_Read_UINT32(s, code);
+		if (code != 0x00000000)
```
better use the `AUTHZ_SUCCESS` and `AUTHZ_ACCESS_DENIED` defines here

hlefebvre replied:
ok changed

akallabeth commented on libfreerdp/core/nla.c:
```diff
@@ -1552,63 +1580,80 @@ int nla_recv_pdu(rdpNla* nla, wStream* s)
+		Stream_Read_UINT32(s, code);
+		if (code != 0x00000000)
+		{
+			WLog_DBG(TAG, "Early User Auth active: FAILURE");
```
maybe log the result code too (should only be 2 allowed, but you never know when that will change)

hlefebvre replied:
I agree with you, code updated

akallabeth reviewed (approved):
LGTM,
nice patch, thank you for your contribution!

hardening commented on libfreerdp/core/transport.c:
```diff
@@ -374,6 +375,59 @@ BOOL transport_connect_nla(rdpTransport* transport)
 	return rdp_client_transition_to_state(rdp, CONNECTION_STATE_NLA);
 }
 
+BOOL transport_connect_nla_ext(rdpTransport* transport)
```
Hum this function is a copy'n paste of `transport_connect_nla`, instead I'd suggest to have a `transport_connect_nla_ex(rdpTransport* transport, BOOL isExt)` and `transport_connect_nla_ext` and `transport_connect_nla` would call this one setting the correct value for `isExt`.

hlefebvre replied:
I agree, I remove the nla_ext function and add a BOOL param to connect_nla should be better

hlefebvre commented:
> LGTM, nice patch, thank you for your contribution!
> 
> if you could squash the commits into one and fix the (minor) remarks it would be really great!

ok I fixed everything and squashed it

hardening commented on libfreerdp/core/nla.c:
```diff
@@ -54,6 +54,9 @@
 
 #define NLA_AUTH_PKG NEGO_SSP_NAME
 
+static const UINT32 AUTHZ_SUCCESS = 0x00000000;
```
nitpicking but in FreeRDP we use enums for these defines

hlefebvre replied:
ok I never used them in C, I've changed it to an enum

hardening commented:
Fix my remark and we're ok (we'll ask for test bots after that)

hardening commented:
@freerdp-bot can you test buddy ?

hlefebvre commented:
I've fixed the wrong indentation for the enum

akallabeth commented on libfreerdp/core/nla.c:
```diff
@@ -54,6 +54,12 @@
 
+typedef enum
+{
+	AUTHZ_SUCCESS = 0x00000000,
```
redefinition, already defined in `nego.h`

akallabeth replied:
ok, pushed a fix.

akallabeth commented:
@freerdp-bot test please

akallabeth commented:
@freerdp-bot test please
